hydra-node/src/Hydra/Ledger/Cardano.hs

```diff
@@ -95,712 +82,510 @@ cardanoLedger =
-  | initialUtxo == mempty = pure []
-  | otherwise = do
-    reverse . snd <$> foldM newTx (initialUtxo, []) [1 .. numTxs]
- where
-  newTx (utxos, acc) _ = do
-    tx <- genCardanoTx utxos
-    case applyTransactions cardanoLedger utxos [tx] of
-      Left err -> error $ show err
-      Right newUtxos -> pure (newUtxos, tx : acc)
+-- | Newtype with phantom types mostly required to work around the poor interface
+-- of 'Ledger.UTXO'and provide 'Monoid' and 'Foldable' instances to make utxo
+-- manipulation bareable.
+newtype Utxo' era out = Utxo
+  { unUtxo :: Map (Ledger.TxIn (Ledger.Crypto era)) out
+  }
+  deriving newtype (Eq, Show)
 
-type CardanoEra = MaryEra StandardCrypto
+instance ToCBOR Utxo where
+  toCBOR = toCBOR . coerce @_ @(Ledger.UTxO LedgerEra)
+  encodedSizeExpr sz _ = encodedSizeExpr sz (Proxy @(Ledger.UTxO LedgerEra))
 
-type CardanoTxBody = Cardano.TxBody CardanoEra
+instance FromCBOR Utxo where
+  fromCBOR = coerce @(Ledger.UTxO LedgerEra) <$> fromCBOR
+  label _ = label (Proxy @(Ledger.UTxO LedgerEra))
 
-type CardanoTxWitnesses = Cardano.WitnessSet CardanoEra
+instance Functor (Utxo' era) where
+  fmap fn (Utxo u) = Utxo (fmap fn u)
 
-type CardanoAuxiliaryData = StrictMaybe (AuxiliaryData CardanoEra)
+instance Foldable (Utxo' era) where
+  foldMap fn = foldMap fn . unUtxo
+  foldr fn zero = foldr fn zero . unUtxo
 
-type CardanoAddress = Cardano.Addr StandardCrypto
+instance Semigroup Utxo where
+  Utxo uL <> Utxo uR = Utxo (uL <> uR)
 
-type CardanoKeyPair = Cardano.KeyPair 'Cardano.Payment StandardCrypto
+instance Monoid Utxo where
+  mempty = Utxo mempty
 
-type TxIn = Cardano.TxIn StandardCrypto
+instance ToJSON Utxo where
+  toJSON = toJSON . unUtxo
 
-type TxOut = Cardano.TxOut CardanoEra
+instance FromJSON Utxo where
+  parseJSON = fmap Utxo . parseJSON
 
-signWith ::
-  Cardano.TxId StandardCrypto ->
-  Cardano.KeyPair 'Cardano.Payment StandardCrypto ->
-  Cardano.WitVKey 'Cardano.Witness StandardCrypto
-signWith (Cardano.TxId safeHash) credentials =
-  Cardano.WitVKey
-    (asWitness $ Cardano.vKey credentials)
-    (signedDSIGN @StandardCrypto (Cardano.sKey credentials) (extractHash safeHash))
+instance Arbitrary Utxo where
+  -- TODO: shrinker!
+  arbitrary = genUtxo
 
---
--- Balance
---
+prettyUtxo :: (TxIn, TxOut ctx era) -> Text
+prettyUtxo (k, TxOut _ (txOutValueToValue -> v) _) =
+  T.drop 54 (renderTxIn k) <> " ↦ " <> prettyValue v
 
-prettyBalance :: Balance tx -> Text
-prettyBalance Balance{lovelace, assets} =
-  let (ada, decimal) = lovelace `quotRem` 1000000
-      padLeft c n str = T.takeEnd n (T.replicate n c <> str)
-   in unwords $
-        [ show ada <> "." <> padLeft "0" 6 (show decimal)
-        , "₳"
+-- FIXME: This function is wrong! It's mapping a transaction's own inputs to its
+-- own outputs. Whoops. It's currently used in Hydra.Chain.Direct.Tx where
+-- the calling code only look at the outputs of the transactions and also in the
+-- generators of the local-cluster (Whoops bis).
```

Comment:
You are right. Would a correct implementation where the `TxId` and `TxIx` are correctly referring to the given `CardanoTx`s outputs be helpful to have in your opinion?